File: dsp_2026_2/06_aliasing.m
% 06_aliasing.m
% Tema 1 - Aliasing: muestreo por debajo de Nyquist (fs < 2 f0).
% Ejecutar:  octave-cli 06_aliasing.m
%

clear; close all; clc;

f0 = 100;                   % frecuencia de la senal [Hz]
t = 0:1e-5:0.1;

% --- Caso 1: muestreo insuficiente (fs < 2 f0) ---------------------------
fs = 70;                    % 70 < 200 Hz -> hay aliasing
Ts = 1/fs;
n = 0:round(0.1/Ts);
tn = n * Ts;

xt = cos(2*pi*f0*t);
xn = cos(2*pi*f0*tn);

% Frecuencia del alias: fa = |f0 - k*fs| (primer alias, k = 1)
fa = abs(f0 - fs);

figure('Name', 'Aliasing');
plot(t, xt, 'b'); hold on;
stem(tn, xn, 'r', 'filled', 'LineWidth', 1.2);
plot(t, cos(2*pi*fa*t), 'g--', 'LineWidth', 1.5);
title(sprintf('Muestreo insuficiente: fs = %g Hz < 2 f0 = %g Hz', fs, 2*f0));
xlabel('t [s]'); ylabel('Amplitud');
legend('x(t) = cos(2\pi 100 t)', ['Muestras fs = ' num2str(fs) ' Hz'], ...
       ['Alias ' num2str(fa) ' Hz'], 'Location', 'northeast');
ylim([-1.3 1.3]);

fprintf('fs = %g Hz < 2*f0 = %g Hz: aparece un alias en %g Hz.\n', fs, 2*f0, fa);

% --- Caso 2: muestreo correcto (fs > 2 f0) --------------------------------
fs2 = 250;                  % 250 > 200 Hz -> sin aliasing
Ts2 = 1/fs2;
n2 = 0:round(0.1/Ts2);
tn2 = n2 * Ts2;
xn2 = cos(2*pi*f0*tn2);

figure('Name', 'Muestreo correcto');
plot(t, xt, 'b'); hold on;
stem(tn2, xn2, 'r', 'filled', 'LineWidth', 1.2);
title(sprintf('Muestreo correcto: fs = %g Hz > 2 f0 = %g Hz', fs2, 2*f0));
xlabel('t [s]'); ylabel('Amplitud');
legend('x(t) = cos(2\pi 100 t)', ['Muestras fs = ' num2str(fs2) ' Hz']);
ylim([-1.3 1.3]);

%% Ejercicio 1
% Verifique la regla del alias: para f0 = 100 Hz pruebe fs = 80, 90, 150, 180
% y calcule fa = abs(f0 - fs) en cada caso. Compruebe graficamente que las
% muestras "caminan" sobre la coseno de frecuencia fa.

%% Ejercicio 2 (propuesto)
% Un audio con fmax = 20 kHz muestreado a fs = 44.1 kHz: verifique que
% fs > 2*fmax. Repita el ejercicio con fs = 32 kHz y explique por que el
% resultado cambia.

File: dsp_2026_2/06_aliasing.py
#!/usr/bin/env python3
"""06_aliasing.py

Tema 1 - Aliasing: muestreo por debajo de Nyquist (fs < 2 f0).
Ejecutar:  python3 06_aliasing.py

Uso de SciPy: scipy.fft.fft para ver el alias en el espectro muestreado.
"""

import matplotlib.pyplot as plt
import numpy as np
from scipy.fft import fft, fftfreq

f0 = 100.0
t = np.arange(0, 0.1, 1e-5)

# --- Caso 1: muestreo insuficiente (fs < 2 f0) ---
fs = 70.0                     # 70 < 200 Hz -> hay aliasing
Ts = 1.0 / fs
n = np.arange(0, int(0.1 / Ts) + 1)
tn = n * Ts

xt = np.cos(2 * np.pi * f0 * t)
xn = np.cos(2 * np.pi * f0 * tn)
fa = abs(f0 - fs)             # primer alias

fig, ax = plt.subplots()
ax.plot(t, xt, "b")
ax.stem(tn, xn, "r", basefmt=" ")
ax.plot(t, np.cos(2 * np.pi * fa * t), "g--", linewidth=1.5)
ax.set_title(f"Muestreo insuficiente: fs = {fs:.0f} Hz < 2 f0 = {2 * f0:.0f} Hz")
ax.set_xlabel("t [s]")
ax.set_ylabel("Amplitud")
ax.legend([r"$x(t)=\cos(2\pi\,100 t)$", f"Muestras fs = {fs:.0f} Hz",
           f"Alias {fa:.0f} Hz"])
ax.set_ylim(-1.3, 1.3)
plt.tight_layout()

print(f"fs = {fs:.0f} Hz < 2*f0 = {2 * f0:.0f} Hz: aparece un alias en {fa:.0f} Hz.")

# Espectro de las muestras (el alias se ve como un pico en fa)
Xk = fft(xn)
freq = fftfreq(len(xn), Ts)
fig2, ax2 = plt.subplots()
ax2.plot(freq[: len(xn) // 2], np.abs(Xk[: len(xn) // 2]))
ax2.set_title("Espectro de las muestras (fs = 70 Hz)")
ax2.set_xlabel("Frecuencia [Hz]")
ax2.set_xlim(0, 120)
plt.tight_layout()

# --- Caso 2: muestreo correcto (fs > 2 f0) ---
fs2 = 250.0
Ts2 = 1.0 / fs2
n2 = np.arange(0, int(0.1 / Ts2) + 1)
tn2 = n2 * Ts2
xn2 = np.cos(2 * np.pi * f0 * tn2)

fig3, ax3 = plt.subplots()
ax3.plot(t, xt, "b")
ax3.stem(tn2, xn2, "r", basefmt=" ")
ax3.set_title(f"Muestreo correcto: fs = {fs2:.0f} Hz > 2 f0 = {2 * f0:.0f} Hz")
ax3.set_xlabel("t [s]")
ax3.set_ylabel("Amplitud")
ax3.legend([r"$x(t)=\cos(2\pi\,100 t)$", f"Muestras fs = {fs2:.0f} Hz"])
ax3.set_ylim(-1.3, 1.3)
plt.tight_layout()

# Ejercicio 1: para f0 = 100 Hz pruebe fs = 80, 90, 150, 180 y calcule
# fa = abs(f0 - fs) en cada caso; compruebe que las muestras "caminan" sobre
# la coseno de frecuencia fa.
# Ejercicio 2 (propuesto): un audio con fmax = 20 kHz muestreado a
# fs = 44.1 kHz cumple fs > 2*fmax. Repita con fs = 32 kHz y explique por qué
# el resultado cambia.

plt.show()
